### PR TITLE
Accessor for path

### PR DESCRIPTION
**Pull request by dinomite**
Suggestions for testing?  Maybe suite-ify a bunch of tests and run them for toUrl() & toUrlString()?

### CODE REVIEW

File: url-builder/src/main/java/com/palominolabs/http/url/UrlBuilder.java [hunk 1]
```diff
@@ -366,11 +355,52 @@ public String toUrlString() throws CharacterCodingException {
...
+    }
 
+    private void appendFragment(StringBuilder buf) throws MalformedInputException, UnmappableCharacterException {
         if (fragment != null) {
             buf.append('#');
             buf.append(fragmentEncoder.encode(fragment));
         }
+    }
+
+    public URL toUrl() throws CharacterCodingException, MalformedURLException {
```
Comment (marshallpierce) [inline url-builder/src/main/java/com/palominolabs/http/url/UrlBuilder.java]: We should probably find a way to not throw MalformedURLException. Wrap in RuntimeException maybe? Or our own type? If we generate a malformed URL that seems like an error in this library

**Comment by dinomite** [comment]
I wrapped the MalformeUrlException in an RTE with the message "Unknown scheme specified", as the Javadoc lists that as the only condition for which a MUE will be thrown for the two constructors we're using.

File: url-builder/src/main/java/com/palominolabs/http/url/UrlBuilder.java [hunk 2]
```diff
@@ -366,11 +352,56 @@ public String toUrlString() throws CharacterCodingException {
...
+    }
 
+    private void appendFragment(StringBuilder buf) throws CharacterCodingException {
         if (fragment != null) {
             buf.append('#');
             buf.append(fragmentEncoder.encode(fragment));
         }
+    }
+
+    public URL toUrl() throws CharacterCodingException {
```
Comment (marshallpierce) [inline url-builder/src/main/java/com/palominolabs/http/url/UrlBuilder.java]: document when it throws RuntimeException

File: url-builder/src/main/java/com/palominolabs/http/url/UrlBuilder.java [hunk 3]
```diff
@@ -366,11 +352,56 @@ public String toUrlString() throws CharacterCodingException {
...
+        String path = buf.toString();
+
+        try {
+            if (port != null) {
+                return new URL(scheme, host, port, path);
+            } else {
+                return new URL(scheme, host, path);
+            }
+        } catch (MalformedURLException e) {
+            throw new RuntimeException("Unknown scheme specified", e);
```
Comment (marshallpierce) [inline url-builder/src/main/java/com/palominolabs/http/url/UrlBuilder.java]: maybe IllegalArgEx just so people can catch () something more specific?

File: url-builder/src/test/java/com/palominolabs/http/url/UrlBuilderTest.java [hunk 4]
```diff
@@ -411,6 +411,26 @@ public void testClearQueryAfterQueryAllowsQueryParam() throws CharacterCodingExc
         assertUrlEquals("http://host?foo=bar", ub.toUrlString());
     }
 
+    @Test
+    public void testToUrlMatchesToUrlString() throws MalformedURLException, CharacterCodingException {
+        UrlBuilder ub = forHost("http", "host")
+                .unstructuredQuery("foobar")
+                .clearQuery()
```
Comment (marshallpierce) [inline url-builder/src/test/java/com/palominolabs/http/url/UrlBuilderTest.java]: no point in adding then clearing a query; copy pasta?

File: url-builder/src/test/java/com/palominolabs/http/url/UrlBuilderTest.java [hunk 5]
```diff
@@ -411,6 +411,26 @@ public void testClearQueryAfterQueryAllowsQueryParam() throws CharacterCodingExc
...
+                .queryParam("foo", "bar");
+        assertEquals(ub.toUrlString(), ub.toUrl().toString());
```
Comment (marshallpierce) [inline url-builder/src/test/java/com/palominolabs/http/url/UrlBuilderTest.java]: let's make this a more full featured url with path and fragment too so we can catch a change that drops passing those to the URL ctor

**Comment by marshallpierce** [comment]
Closing; idle. Re-open if needed.

**Pull request closed without merging**